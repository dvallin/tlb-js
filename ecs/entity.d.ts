import { World } from './world';
export declare type Entity = number;
export declare class EntityModifier<C, S, R> {
    private readonly world;
    readonly entity: Entity;
    constructor(world: World<C, S, R>, entity: Entity);
    withComponent<T extends object>(name: C, component: T): EntityModifier<C, S, R>;
    removeComponent(name: C): EntityModifier<C, S, R>;
    delete(): void;
}
