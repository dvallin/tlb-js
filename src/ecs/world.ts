import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
