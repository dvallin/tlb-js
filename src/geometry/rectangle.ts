import { AbstractShape } from "./shape"
import { Vector } from "../spatial"

export class Rectangle extends AbstractShape {

    public static fromBounds(left: number, right: number, top: number, bottom: number): Rectangle {
        return new Rectangle(left, top, right - left + 1, bottom - top + 1)
    }

    public constructor(
        public readonly x: number,
        public readonly y: number,
        public readonly width: number,
        public readonly height: number,
    ) {
        super()
    }

    public get left(): number {
        return this.x
    }

    public get right(): number {
        return this.x + this.width - 1
    }

    public get top(): number {
        return this.y
    }

    public get bottom(): number {
        return this.y + this.height - 1
    }

    public get topLeft(): Vector {
        return new Vector(this.left, this.top)
    }

    public get topRight(): Vector {
        return new Vector(this.right, this.top)
    }

    public get bottomRight(): Vector {
        return new Vector(this.right, this.bottom)
    }

    public get bottomLeft(): Vector {
        return new Vector(this.left, this.bottom)
    }

    public get center(): Vector {
        const cx = (this.right + this.left) / 2
        const cy = (this.bottom + this.top) / 2
        return new Vector(Math.floor(cx), Math.floor(cy))
    }

    public plus(other: Rectangle): Rectangle {
        return Rectangle.fromBounds(
            Math.min(this.left, other.left), Math.max(this.right, other.right),
            Math.min(this.top, other.top), Math.max(this.bottom, other.bottom)
        )
    }

    public bounds(): Rectangle {
        return this
    }

    public containsVector(p: Vector): boolean {
        return p.x >= this.x && p.x < this.x + this.width
            && p.y >= this.y && p.y < this.y + this.height
    }

    public translate(t: Vector): Rectangle {
        return new Rectangle(this.x + t.x, this.y + t.y, this.width, this.height)
    }

    public grow(): Rectangle {
        return new Rectangle(this.x - 1, this.y - 1, this.width + 2, this.height + 2)
    }

    public all(f: (position: Vector) => boolean): boolean {
        for (let i = 0; i < this.height; i++) {
            for (let j = 0; j < this.width; j++) {
                const position = new Vector(j + this.x, i + this.y)
                if (!f(position)) {
                    return false
                }
            }
        }
        return true
    }
}
