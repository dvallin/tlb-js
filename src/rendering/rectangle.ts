import { Rectangle } from "@/geometry/Rectangle"
import { Position } from "@/geometry/Position"
import { RenderIterator } from "@/rendering"

export function rasterize(rectangle: Rectangle, fill: boolean = false): RenderIterator {
    let y = rectangle.top
    let x = rectangle.left
    let done = false

    return () => {
        if (done) {
            return undefined
        }
        const result = new Position(Math.floor(x), Math.floor(y))
        x++
        if (x > rectangle.right) {
            x = rectangle.left
            y++
        }
        if (!fill && x > rectangle.left && y !== rectangle.top && y !== rectangle.bottom) {
            x = rectangle.right
        }
        if (y > rectangle.bottom) {
            done = true
        }
        return result
    }
}
