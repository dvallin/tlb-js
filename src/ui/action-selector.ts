import { Action, SelectedAction } from '../components/action'
import { Rectangle } from '../geometry/rectangle'
import { Entity } from '../ecs/entity'
import { Vector } from '../spatial'
import { Renderer } from '../renderer/renderer'
import { Difference } from '../geometry/difference'
import { primary, gray } from '../renderer/palettes'

export interface SelectableAction {
  action: Action
  available: boolean
}

export interface ActionGroup {
  entity: number
  description: string
  name: string
  actions: SelectableAction[]
}

export interface State {
  title: string
  actionsWindow: Rectangle
  descriptionWindow: Rectangle
  element: Entity
  rows: number

  groups: ActionGroup[]

  selected: number | undefined
  hovered: number
}

export class ActionSelector {
  public constructor(private readonly state: State) {}

  public render(renderer: Renderer) {
    const { actionsWindow, descriptionWindow, groups, hovered } = this.state
    this.renderActions(renderer, actionsWindow, groups, hovered)
    this.renderDescription(renderer, descriptionWindow, groups, hovered)
  }

  public get element(): Entity {
    return this.state.element
  }

  public update(position: Vector | undefined, pressed: boolean, up: boolean, down: boolean) {
    this.state.selected = undefined
    const content = this.state.actionsWindow.shrink()
    if (position && content.containsVector(position)) {
      const delta = position.minus(content.topLeft)
      this.state.hovered = delta.y
    }
    if (up) {
      this.state.hovered--
    }
    if (down) {
      this.state.hovered++
    }
    this.state.hovered += this.state.rows
    this.state.hovered %= this.state.rows

    if (pressed) {
      this.state.selected = this.state.hovered
    }
  }

  public get selectedAction(): SelectedAction | undefined {
    if (this.state.selected !== undefined) {
      const { groups, selected } = this.state
      return this.actionAtLine(groups, selected)
    }
    return undefined
  }

  private renderActions(renderer: Renderer, window: Rectangle, groups: ActionGroup[], hovered: number) {
    this.renderWindowBorder(renderer, window, 'actions')

    let index = 0
    groups.forEach(group => {
      renderer.text(`- ${group.name}`, window.topLeft.add(new Vector(1, index + 1)), primary[0], hovered === index ? gray[1] : undefined)
      index++
      group.actions.forEach(action => {
        renderer.text(
          ` | ${action.action.name}`,
          window.topLeft.add(new Vector(1, index + 1)),
          primary[0],
          hovered === index ? gray[1] : undefined
        )
        index++
      })
    })
  }

  private renderDescription(renderer: Renderer, window: Rectangle, groups: ActionGroup[], hovered: number) {
    this.renderWindowBorder(renderer, window, 'description')
    const action = this.actionAtLine(groups, hovered)
    if (action !== undefined) {
      renderer.text(action.action.name, window.topLeft.add(new Vector(1, 1)), primary[0])
      const cost = action.action.cost
      if (cost.costsAll) {
        renderer.text('costs all AP and MP', window.topLeft.add(new Vector(1, 2)), primary[0])
      } else {
        renderer.text(`cost: ${cost.actions}AP ${cost.movement}MP`, window.topLeft.add(new Vector(1, 2)), primary[0])
      }
    } else {
      const group = this.groupAtLine(groups, hovered)
      if (group !== undefined) {
        renderer.text(group.name, window.topLeft.add(new Vector(1, 1)), primary[0])
        renderer.flowText(group.description, window.topLeft.add(new Vector(1, 2)), window.width - 2, primary[0])
      }
    }
  }

  private renderWindowBorder(renderer: Renderer, window: Rectangle, title: string) {
    Difference.innerBorder(window).foreach(p => {
      if (p.y === window.top || p.y === window.bottom) {
        renderer.character('-', p, primary[0])
      } else {
        renderer.character('|', p, primary[0])
      }
    })
    const titleText = `/${title}/`
    renderer.text(titleText, { x: window.right - titleText.length, y: window.top }, primary[0])
  }

  private groupAtLine(groups: ActionGroup[], line: number): ActionGroup | undefined {
    for (const group of groups) {
      if (line === 0) {
        return group
      }
      line -= group.actions.length + 1
    }
    return undefined
  }

  private actionAtLine(groups: ActionGroup[], line: number): SelectedAction | undefined {
    for (const group of groups) {
      line -= 1
      if (line < 0) {
        return undefined
      } else if (line >= group.actions.length) {
        line -= group.actions.length
      } else {
        return {
          entity: group.entity,
          action: group.actions[line].action,
        }
      }
    }
    return undefined
  }

  public hasElement(position: Vector): boolean {
    return this.state.actionsWindow!.containsVector(position) || this.state.descriptionWindow!.containsVector(position)
  }
}
