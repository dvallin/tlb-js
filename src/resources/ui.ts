import { ResourceName, TlbResource, TlbWorld } from '../tlb'
import { Vector } from '../spatial'
import { Entity } from '../ecs/entity'
import { Difference } from '../geometry/difference'
import { Rectangle } from '../geometry/rectangle'
import { Renderer } from '../renderer/renderer'
import { primary, gray } from '../renderer/palettes'
import { InputResource, Input } from './input'
import { ViewportResource } from './viewport'

export interface UI {
  hasElement(position: Vector): boolean
  render(renderer: Renderer): void

  showSelectList(world: TlbWorld, entries: string[]): void
  selectListSelection(): string | undefined
  hideSelectList(world: TlbWorld): void
}

export interface SelectList {
  entries: string[]
  shape: Rectangle
  element: Entity
  hovered: number | undefined
  selected: number | undefined
}

export class UIResource implements TlbResource, UI {
  public readonly kind: ResourceName = 'ui'

  public visibleElements: Rectangle[] = []

  private selectList: SelectList | undefined = undefined

  public update(world: TlbWorld): void {
    const input: Input = world.getResource<InputResource>('input')
    let position
    if (input.position) {
      position = new Vector(input.position.x, input.position.y)
    }
    this.updateSelectList(position, input.mousePressed)
  }

  public updateSelectList(position: Vector | undefined, pressed: boolean) {
    if (this.selectList !== undefined) {
      this.selectList.hovered = undefined
      this.selectList.selected = undefined
      const content = this.selectList.shape.shrink()
      if (position && content.containsVector(position)) {
        const delta = position.minus(content.topLeft)
        this.selectList.hovered = delta.y
        if (pressed) {
          this.selectList.selected = delta.y
        }
      }
    }
  }

  public render(renderer: Renderer): void {
    this.renderInfoPopup(renderer)
  }

  public renderInfoPopup(renderer: Renderer) {
    if (this.selectList !== undefined) {
      const list = this.selectList
      Difference.innerBorder(list.shape).foreach(p => {
        renderer.character('+', p, primary[0])
      })
      list.entries.forEach((action, index) => {
        renderer.text(action, list.shape.topLeft.add(new Vector(1, index + 1)), primary[0], list.hovered === index ? gray[1] : undefined)
      })
    }
  }

  public showSelectList(world: TlbWorld, entries: string[]) {
    const viewport = world.getResource<ViewportResource>('viewport')
    const height = entries.length + 2
    let element
    if (this.selectList !== undefined) {
      element = this.selectList.element
    } else {
      element = world.createEntity().entity
    }
    const shape = new Rectangle(30, viewport.boundaries.y - height, 20, height)
    this.selectList = { entries, shape, element, hovered: undefined, selected: undefined }
  }

  public hideSelectList(world: TlbWorld) {
    if (this.selectList !== undefined) {
      world.deleteEntity(this.selectList.element)
      this.selectList = undefined
    }
  }

  public selectListSelection(): string | undefined {
    if (this.selectList !== undefined && this.selectList.selected !== undefined) {
      return this.selectList.entries[this.selectList.selected]
    }
    return undefined
  }

  public hasElement(position: Vector): boolean {
    return this.selectList !== undefined && this.selectList.shape!.containsVector(position)
  }
}
