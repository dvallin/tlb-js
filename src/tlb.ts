import { World } from './ecs/world'
import { VectorStorage, MapStorage, SetStorage, SingletonStorage } from './ecs/storage'
import { System } from './ecs/system'
import { Resource } from './ecs/resource'

import { AgeComponent } from './components/age'
import { AssetComponent } from './components/asset'
import { CharacterStatsComponent } from './components/character-stats'
import { FeatureComponent } from './components/feature'
import { FovComponent } from './components/fov'
import { GroundComponent } from './components/ground'
import { PositionComponent } from './components/position'
import { RegionComponent, StructureComponent } from './components/region'

import { Fov } from './systems/fov'
import { FreeModeControl } from './systems/free-mode-control'
import { Npc } from './systems/npc'
import { PlayerControl } from './systems/player-control'
import { PlayerInteraction } from './systems/player-interaction'
import { RegionCreator } from './systems/region-creator'
import { PlayerRoundControl } from './systems/player-round-control'
import { Trigger } from './systems/trigger'
import { ApplyEffects } from './systems/apply-effect'
import { Script } from './systems/script'

import { WorldMapResource } from './resources/world-map'
import { ViewportResource } from './resources/viewport'
import { InputResource } from './resources/input'
import { UIResource } from './resources/ui'
import { LogResource } from './resources/log'

import { Random } from './random'
import { Uniform } from './random/distributions'

import { Renderer } from './renderer/renderer'
import { Queries } from './renderer/queries'
import { State } from './game-states/state'
import { OverlayComponent } from './components/overlay'
import { TakeTurnComponent } from './components/rounds'
import { SelectedActionComponent, HasActionComponent } from './components/action'
import { ScriptComponent } from './components/script'
import { AiComponent } from './components/ai'
import { AiRoundControl } from './systems/ai-round-control'
import { TriggersComponent, TriggeredByComponent } from './components/trigger'
import { EffectComponent, ActiveEffectsComponent } from './components/effects'
import { InventoryComponent, ItemComponent, EquipedItemsComponent } from './components/items'
import { StartRound } from './systems/start-round'
import { DialogComponent } from './components/dialog'
export type ComponentName =
  | 'active'
  | 'age'
  | 'ai'
  | 'asset'
  | 'character-stats'
  | 'dead'
  | 'dialog'
  | 'effect'
  | 'equiped-items'
  | 'feature'
  | 'fov'
  | 'free-mode-anchor'
  | 'ground'
  | 'has-action'
  | 'inventory'
  | 'item'
  | 'npc'
  | 'overlay'
  | 'player'
  | 'position'
  | 'region'
  | 'structure'
  | 'script'
  | 'selected-action'
  | 'active-effects'
  | 'spawn'
  | 'start-turn'
  | 'take-turn'
  | 'took-turn'
  | 'triggered-by'
  | 'triggers'
  | 'viewport-focus'
  | 'wait-turn'
export type SystemName =
  | 'ai-round-control'
  | 'start-round'
  | 'effect'
  | 'fov'
  | 'free-mode-control'
  | 'info-popup'
  | 'npc'
  | 'player-control'
  | 'player-interaction'
  | 'player-round-control'
  | 'region-creator'
  | 'script'
  | 'trigger'
export type ResourceName = 'input' | 'map' | 'viewport' | 'ui' | 'log'

export type TlbWorld = World<ComponentName, SystemName, ResourceName>
export type TlbResource = Resource<ComponentName, SystemName, ResourceName>
export type TlbSystem = System<ComponentName, SystemName, ResourceName>

export function registerComponents<S, R>(world: World<ComponentName, S, R>): void {
  world.registerComponentStorage('active', new SetStorage())
  world.registerComponentStorage('age', new MapStorage<AgeComponent>())
  world.registerComponentStorage('ai', new MapStorage<AiComponent>())
  world.registerComponentStorage('asset', new MapStorage<AssetComponent>())
  world.registerComponentStorage('character-stats', new MapStorage<CharacterStatsComponent>())
  world.registerComponentStorage('dead', new SetStorage())
  world.registerComponentStorage('dialog', new MapStorage<DialogComponent>())
  world.registerComponentStorage('effect', new MapStorage<EffectComponent>())
  world.registerComponentStorage('feature', new VectorStorage<FeatureComponent>())
  world.registerComponentStorage('fov', new MapStorage<FovComponent>())
  world.registerComponentStorage('free-mode-anchor', new SingletonStorage<{}>())
  world.registerComponentStorage('ground', new MapStorage<GroundComponent>())
  world.registerComponentStorage('has-action', new MapStorage<HasActionComponent>())
  world.registerComponentStorage('inventory', new MapStorage<InventoryComponent>())
  world.registerComponentStorage('structure', new MapStorage<StructureComponent>())
  world.registerComponentStorage('equiped-items', new MapStorage<EquipedItemsComponent>())
  world.registerComponentStorage('item', new MapStorage<ItemComponent>())
  world.registerComponentStorage('npc', new SetStorage())
  world.registerComponentStorage('overlay', new MapStorage<OverlayComponent>())
  world.registerComponentStorage('player', new SingletonStorage<{}>())
  world.registerComponentStorage('position', new VectorStorage<PositionComponent>())
  world.registerComponentStorage('region', new MapStorage<RegionComponent>())
  world.registerComponentStorage('script', new MapStorage<ScriptComponent>())
  world.registerComponentStorage('selected-action', new SingletonStorage<SelectedActionComponent>())
  world.registerComponentStorage('active-effects', new MapStorage<ActiveEffectsComponent>())
  world.registerComponentStorage('spawn', new SingletonStorage<{}>())
  world.registerComponentStorage('take-turn', new MapStorage<TakeTurnComponent>())
  world.registerComponentStorage('took-turn', new SetStorage())
  world.registerComponentStorage('start-turn', new SetStorage())
  world.registerComponentStorage('triggered-by', new MapStorage<TriggeredByComponent>())
  world.registerComponentStorage('triggers', new MapStorage<TriggersComponent>())
  world.registerComponentStorage('viewport-focus', new SingletonStorage<{}>())
  world.registerComponentStorage('wait-turn', new SetStorage())
}

export function registerResources(
  world: World<ComponentName, SystemName, ResourceName>,
  renderer: Renderer,
  worldWidth: number = 256
): void {
  world.registerResource(new WorldMapResource(worldWidth))
  world.registerResource(new ViewportResource(renderer.boundaries))
  world.registerResource(new InputResource(e => renderer.eventToPosition(e)))
  world.registerResource(new UIResource())
  world.registerResource(new LogResource())
}

export function registerSystems(
  world: World<ComponentName, SystemName, ResourceName>,
  queries: Queries,
  pushState: (s: State) => void
): void {
  const uniform = new Uniform('some seed')
  world.registerSystem('ai-round-control', new AiRoundControl(queries, new Random(uniform)))
  world.registerSystem('effect', new ApplyEffects())
  world.registerSystem('fov', new Fov(queries))
  world.registerSystem('free-mode-control', new FreeModeControl())
  world.registerSystem('npc', new Npc(queries, new Random(uniform), pushState))
  world.registerSystem('player-control', new PlayerControl())
  world.registerSystem('player-interaction', new PlayerInteraction(pushState))
  world.registerSystem('player-round-control', new PlayerRoundControl(queries, new Random(uniform)))
  world.registerSystem('region-creator', new RegionCreator(uniform))
  world.registerSystem('script', new Script())
  world.registerSystem('trigger', new Trigger(new Random(uniform), pushState))
  world.registerSystem('start-round', new StartRound(new Random(uniform)))
}
