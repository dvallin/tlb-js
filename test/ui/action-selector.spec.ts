import { ActionSelector, ActionGroup } from '../../src/ui/action-selector'
import { Rectangle } from '../../src/geometry/rectangle'
import { World } from '../../src/ecs/world'
import { TlbWorld } from '../../src/tlb'
import { mockInput } from '../mocks'
import { KEYS } from 'rot-js'

describe('ActionSelector', () => {
  describe('two uncollapsed groups', () => {
    const group1: ActionGroup = createGroup('1', '1', '2', '3')
    const group2: ActionGroup = createGroup('2', '1', '2')

    describe('hover', () => {
      let selector: ActionSelector
      beforeEach(() => {
        selector = ActionSelector.build(0, new Rectangle(0, 0, 10, 5), [group1, group2])
      })

      it('sets hovered by mouse position', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.position = { x: 3, y: 3 }

        selector.update(world)

        expect(selector.hovered).toEqual({
          action: group1.items[1].action,
          entity: group1.entity,
        })
      })

      it('hovers first row on initial down key', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)

        expect(selector.groupSelector.hovered).toBeUndefined()
        expect(selector.hovered).toEqual({
          action: group1.items[0].action,
          entity: group1.entity,
        })
      })

      it('hovers second row on subsequent down key', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)

        expect(selector.hovered).toEqual({
          action: group1.items[1].action,
          entity: group1.entity,
        })
      })

      it('hovers first group initially', () => {
        const world: TlbWorld = new World()
        mockInput(world)

        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group1, collapsed: false })
        expect(selector.hovered).toBeUndefined()
      })

      it('hovers second group', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group2, collapsed: false })
        expect(selector.actionSelector.hovered).toBeUndefined()
      })

      it('hovers first group on wrapping around', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group1, collapsed: false })
        expect(selector.actionSelector.hovered).toBeUndefined()
      })
    })
  })

  describe('two uncollapsed and one collapsed groups', () => {
    const group1: ActionGroup = createGroup('1', '1', '2', '3')
    const group2: ActionGroup = createGroup('2', '1', '2')
    const group3: ActionGroup = createGroup('3', '1', '2')

    describe('hover', () => {
      let selector: ActionSelector
      beforeEach(() => {
        selector = ActionSelector.build(0, new Rectangle(0, 0, 10, 5), [group1, group2, group3])
        clickAt(selector, 4)
      })

      it('sets hovered by mouse position', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.position = { x: 3, y: 3 }

        selector.update(world)

        expect(selector.groupSelector.hovered).toBeUndefined()
        expect(selector.hovered).toEqual({
          action: group1.items[1].action,
          entity: group1.entity,
        })
      })

      it('hovers first row on initial down key', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)

        expect(selector.groupSelector.hovered).toBeUndefined()
        expect(selector.hovered).toEqual({
          action: group1.items[0].action,
          entity: group1.entity,
        })
      })

      it('hovers second row on subsequent down key', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)

        expect(selector.groupSelector.hovered).toBeUndefined()
        expect(selector.hovered).toEqual({
          action: group1.items[1].action,
          entity: group1.entity,
        })
      })

      it('hovers first group initially', () => {
        const world: TlbWorld = new World()
        mockInput(world)

        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group1, collapsed: false })
        expect(selector.actionSelector.hovered).toBeUndefined()
      })

      it('hovers second group', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group2, collapsed: true })
        expect(selector.actionSelector.hovered).toBeUndefined()
      })

      it('hovers third group', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group3, collapsed: false })
        expect(selector.actionSelector.hovered).toBeUndefined()
      })

      it('hovers first group on wrapping around', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group1, collapsed: false })
        expect(selector.actionSelector.hovered).toBeUndefined()
      })
    })

    describe('select', () => {
      let selector: ActionSelector
      beforeEach(() => {
        selector = ActionSelector.build(0, new Rectangle(0, 0, 10, 5), [group1, group2, group3])
        clickAt(selector, 4)
      })

      it('selects by mouse press', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.position = { x: 3, y: 3 }
        input.mousePressed = true

        selector.update(world)

        expect(selector.groupSelector.selected).toBeUndefined()
        expect(selector.selected).toEqual({
          action: group1.items[1].action,
          entity: group1.entity,
        })
      })

      it('selects first row on initial down key', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)
        input.keyPressed.add(KEYS.VK_RETURN)

        selector.update(world)

        expect(selector.groupSelector.selected).toBeUndefined()
        expect(selector.selected).toEqual({
          action: group1.items[0].action,
          entity: group1.entity,
        })
      })

      it('selects second row on subsequent down key', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)
        input.keyPressed.add(KEYS.VK_RETURN)

        selector.update(world)
        selector.update(world)

        expect(selector.groupSelector.selected).toBeUndefined()
        expect(selector.selected).toEqual({
          action: group1.items[1].action,
          entity: group1.entity,
        })
      })

      it('collapses first group initially', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_RETURN)

        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group1, collapsed: true })
        expect(selector.actionSelector.selected).toBeUndefined()
      })

      it('uncollapses second group', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)
        selector.update(world)

        input.keyPressed.add(KEYS.VK_RETURN)
        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group2, collapsed: false })
        expect(selector.actionSelector.selected).toBeUndefined()
      })

      it('collapses third group', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)

        input.keyPressed.add(KEYS.VK_RETURN)
        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group3, collapsed: true })
        expect(selector.actionSelector.hovered).toBeUndefined()
      })

      it('collapses first group on wrapping around', () => {
        const world: TlbWorld = new World()
        const input = mockInput(world)
        input.keyPressed.add(KEYS.VK_J)

        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)
        selector.update(world)

        input.keyPressed.add(KEYS.VK_RETURN)
        selector.update(world)

        expect(selector.groupSelector.hovered).toEqual({ ...group1, collapsed: true })
        expect(selector.actionSelector.hovered).toBeUndefined()
      })
    })
  })
})

function createGroup(id: string, ...actions: string[]): ActionGroup {
  return {
    entity: 1,
    description: `group${id}/description`,
    name: `group${id}/name1`,
    items: actions.map(actionId => ({
      action: { cost: { actions: 1, movement: 2 }, name: `group${id}/actions/${actionId}`, subActions: [] },
      available: true,
    })),
  }
}

function clickAt(selector: ActionSelector, index: number) {
  const world: TlbWorld = new World()
  const input = mockInput(world)
  input.keyPressed.add(KEYS.VK_J)
  for (let i = 0; i < index; i++) {
    selector.update(world)
  }
  expect(selector.groupSelector.hovered).toBeDefined()
  input.keyPressed.delete(KEYS.VK_J)
  input.keyPressed.add(KEYS.VK_RETURN)
  selector.update(world)
  input.keyPressed.delete(KEYS.VK_RETURN)
  input.keyPressed.add(KEYS.VK_K)
  for (let i = 0; i < index; i++) {
    selector.update(world)
  }
}
