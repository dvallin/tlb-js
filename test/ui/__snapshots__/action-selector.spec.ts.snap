// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSelector two uncollapsed and one collapsed groups hover renders actions from the top 1`] = `
"---------------/actions/-----------/description/-
|- group1/name1         |group1/name1           |
| | group1/actions/1    |                       |
| | group1/actions/2    |                       |
| | group1/actions/3    |                       |
|+ group2/name1         |                       |
|- group3/name1         |                       |
-------------------------------------------------
"
`;

exports[`ActionSelector two uncollapsed and one collapsed groups hover scrolls down 1`] = `
"---------------/actions/-----------/description/-
| | group1/actions/1    |group3/actions/1       |
| | group1/actions/2    |cost: 1AP 2MP          |
| | group1/actions/3    |                       |
|+ group2/name1         |                       |
|- group3/name1         |                       |
| | group3/actions/1    |                       |
-------------------------------------------------
"
`;
